### ISSUE TITLE
Advanced Simulation Mode

### ISSUE THREAD

**Issue by RI0** [evil-mad/robopaint#157]
Most of our thinking so far seems to be about RoboPaint connected to a WaterColorBot.

I was thinking about how my class could download RoboPaint at home or on lots of computers in the computer lab. The drawings could be made individually and saved to files ( like the MakerBot )

If the files were brought together, everyone could watch the the bot create the paintings as a group and it would be more fun. 

To help people prepare the files, maybe RoboPaint could have some kind of advanced simulation mode that would allow us to create a simulated preview and then print the drawing to a normal Ink Jet printer as a rough draft.

**Comment by oskay** [comment]
I don't think that it's important -- or even necessarily, a good idea -- to be particularly realistic. In RoboPaint RT (IIRC), we just use a constant-width line, with opacity that fades as a function of distance since the last re-inking. It gives *an idea* of what the output will look like. It's a bit like our "simulation" in scratch. Extra neat, when it will show what all the fills look like.

**Comment by RI0** [comment]
@techninja - I will try to explain the virtual or simulation idea in more detail: 
My starting point is a classroom or group where several people have computers but there is only one robot. Many people will be competing for time with the robot so my idea is to do the following:

The first time someone starts up RoboPaint and no robot is attached, the user is asked if he would like to build a virtual WaterColorBot or a virtual EggBot. 

The user will see a pile of components and then drag the parts into the right places a bit like a puzzle. The pieces will snap into place. If the X motor is in the y motor spot the robot will not work properly but it will work. This could be used as a way of letting everyone build the robots - not as much fun - but better than not doing it. To help make the robots unique and to get the full build experience, I think it would be fun to ask for a number between 1 and 10. This number could be multiplied by a random number between 0.1 and 3.0. The result is the error in the robot. This is the number of steps of error that will exist for this robot. The robot can be saved and the user can give it a name. The robot cannot be changed unless the robot is rebuilt. Maybe the builder will choose 1 next time and get one that is closer to error free. ( not sure 0 should be possible )  

After the virtual robot is built then it can be used in the same way that a real robot can be used. 

RoboPaint will treat it like a real robot. If a painting is sent to the virtual printer, the puzzle pieces come to life and robopaint the drawing in real time. Maybe there could be a turbo button to make it paint really fast but the little carriage must move around and follow the same instructions as the real robot. When the virtual robot is done it will make a little beep to indicate it is finished and then move back to home position. The following options become available : 

Save the painting as JPG / PNG / PDF
Save the painting as an animated GIF 
Save the painting as a set of instructions that can be loaded in to the real robot 
Print the painting using a standard printer

The last option would use the default printer driver and could be used to create draft print outs before using the robot. 

It seems like a lot of things can be done with RoboPaint even if using a virtual or shared WaterColor Robot or an EggBot Robot. This might also allow more people to experiment with the EggBot and create designs that could be shared.

**Comment by RI0** [comment]
@techninja - I agree with your comments. I think this would probably be used for classrooms where there are 20 computers sharing 1 robot. It would allow people to do something instead of just waiting. It could also be used by classes who will probably never get a robot but would still like to do a RoboPaint project with their virtual robots.
